File: src/ImpuestosBolivianos/Invoice.vb
Public Structure Invoice
    Public NroAutorizacion As Int64
    Public NroFactura As Int64
    Public Fecha As Nullable(Of DateTime)
    Public LlaveDosificacion As String
    Public CodigoControl As String
    Public NitEmisor As String
    Public NitCliente As String
    Public ImporteTotal As Decimal
    Public ImporteBaseCf As Decimal
    Public ImporteIceIehdTasas As Decimal
    Public ImporteVentasNoGravadas As Decimal
    Public ImporteNoSujetoCf As Decimal
    Public DescuentosBonosRebajas As Decimal

    Public Sub AssertHasEnoughInfoToMakeCodigoControl()
        ThrowIfInvalid(NroAutorizacion, NameOf(NroAutorizacion))
        ThrowIfInvalid(NroFactura, NameOf(NroFactura))
        ThrowIfInvalid(NitCliente, NameOf(NitCliente))
        ThrowIfInvalid(Fecha, NameOf(Fecha))
        ThrowIfInvalid(ImporteTotal, NameOf(ImporteTotal))
        ThrowIfInvalid(LlaveDosificacion, NameOf(LlaveDosificacion))
    End Sub

    Public Sub AssertHasEnoughInfoToRenderQrCode()
        ThrowIfInvalid(NitEmisor, NameOf(NitEmisor))
        ThrowIfInvalid(NroFactura, NameOf(NroFactura))
        ThrowIfInvalid(NroAutorizacion, NameOf(NroAutorizacion))
        ThrowIfInvalid(Fecha, NameOf(Fecha))
        ThrowIfInvalid(ImporteTotal, NameOf(ImporteTotal))
        ThrowIfInvalid(ImporteBaseCf, NameOf(ImporteBaseCf))
        ThrowIfInvalid(CodigoControl, NameOf(CodigoControl))
        ThrowIfInvalid(NitCliente, NameOf(NitCliente))
        ThrowIfInvalid(ImporteIceIehdTasas, NameOf(ImporteIceIehdTasas))
        ThrowIfInvalid(ImporteVentasNoGravadas, NameOf(ImporteVentasNoGravadas))
        ThrowIfInvalid(ImporteNoSujetoCf, NameOf(ImporteNoSujetoCf))
        ThrowIfInvalid(DescuentosBonosRebajas, NameOf(DescuentosBonosRebajas))
    End Sub

    Private Sub ThrowIfInvalid(value As Int64, argumentName As String)
        If value = 0 Then ThrowArgumentException(argumentName)
    End Sub

    Private Sub ThrowIfInvalid(value As String, argumentName As String)
        If value = String.Empty Then ThrowArgumentException(argumentName)
    End Sub

    Private Sub ThrowIfInvalid(value As Nullable(Of DateTime), argumentName As String)
        If Not value.HasValue Then ThrowArgumentException(argumentName)
    End Sub

    Private Sub ThrowIfInvalid(value As Decimal, argumentName As String)
        If argumentName <> NameOf(ImporteTotal) Then Return

        If value = 0 Then ThrowArgumentException(argumentName)
    End Sub

    Private Sub ThrowArgumentException(argumentName As String)
        Throw New ArgumentException($"{argumentName} is required for this operation.")
    End Sub
End Structure

File: src/ImpuestosBolivianos/Invoice.cs
using System;

namespace ImpuestosBolivianos
{
    public struct Invoice
    {
        public long NroAutorizacion;
        public long NroFactura;
        public DateTime? Fecha;
        public string LlaveDosificacion;
        public string CodigoControl;
        public string NitEmisor;
        public string NitCliente;
        public decimal ImporteTotal;
        public decimal ImporteBaseCf;
        public decimal ImporteIceIehdTasas;
        public decimal ImporteVentasNoGravadas;
        public decimal ImporteNoSujetoCf;
        public decimal DescuentosBonosRebajas;

        public void AssertHasEnoughInfoToMakeCodigoControl()
        {
            ThrowIfInvalid(NroAutorizacion, nameof(NroAutorizacion));
            ThrowIfInvalid(NroFactura, nameof(NroFactura));
            ThrowIfInvalid(NitCliente, nameof(NitCliente));
            ThrowIfInvalid(Fecha, nameof(Fecha));
            ThrowIfInvalid(ImporteTotal, nameof(ImporteTotal));
            ThrowIfInvalid(LlaveDosificacion, nameof(LlaveDosificacion));
        }

        public void AssertHasEnoughInfoToRenderQrCode()
        {
            ThrowIfInvalid(NitEmisor, nameof(NitEmisor));
            ThrowIfInvalid(NroFactura, nameof(NroFactura));
            ThrowIfInvalid(NroAutorizacion, nameof(NroAutorizacion));
            ThrowIfInvalid(Fecha, nameof(Fecha));
            ThrowIfInvalid(ImporteTotal, nameof(ImporteTotal));
            ThrowIfInvalid(ImporteBaseCf, nameof(ImporteBaseCf));
            ThrowIfInvalid(CodigoControl, nameof(CodigoControl));
            ThrowIfInvalid(NitCliente, nameof(NitCliente));
            ThrowIfInvalid(ImporteIceIehdTasas, nameof(ImporteIceIehdTasas));
            ThrowIfInvalid(ImporteVentasNoGravadas, nameof(ImporteVentasNoGravadas));
            ThrowIfInvalid(ImporteNoSujetoCf, nameof(ImporteNoSujetoCf));
            ThrowIfInvalid(DescuentosBonosRebajas, nameof(DescuentosBonosRebajas));
        }

        private void ThrowIfInvalid(long value, string argumentName)
        {
            if (value == 0L)
                ThrowArgumentException(argumentName);
        }

        private void ThrowIfInvalid(string value, string argumentName)
        {
            if ((value ?? "") == (string.Empty ?? ""))
                ThrowArgumentException(argumentName);
        }

        private void ThrowIfInvalid(DateTime? value, string argumentName)
        {
            if (!value.HasValue)
                ThrowArgumentException(argumentName);
        }

        private void ThrowIfInvalid(decimal value, string argumentName)
        {
            if ((argumentName ?? "") != nameof(ImporteTotal))
                return;
            if (value == 0m)
                ThrowArgumentException(argumentName);
        }

        private void ThrowArgumentException(string argumentName)
        {
            throw new ArgumentException($"{argumentName} is required for this operation.");
        }
    }
}